src/context.ts

```diff
@@ -6,12 +6,20 @@ import {
   useMemo,
   useRef,
 } from 'react'
-import { EqualityChecker, State, StateSelector, UseBoundStore } from 'zustand'
+import {
+  EqualityChecker,
+  State,
+  StateSelector,
+  StoreApi,
+  UseBoundStore,
+} from 'zustand'
 
-export type UseContextStore<T extends State> = {
-  (): T
-  <U>(selector: StateSelector<T, U>, equalityFn?: EqualityChecker<U>): U
-}
+export type UseContextStore<T extends State | StoreApi<State>> = <
```

Comment:
We want to mark the old usage with `@deprecated` with a suggestion message. Would it be possible?

Reply:
It is possible but then even if they pass `UseContextStore<MyStore>` (which is the correct newer usage) they'll still see the strikethrough, so they won't be able to distinguish between the types that need to migrated and the types that have already been migrated.

Reply:
that's not nice. ideally, they should see deprecated mark for the old usage and not for the new usage.

Reply:
Yep but we can't have that, the only way to achieve that is have a seperate type but then we'll lose the original name.

Reply:
That's unfortunate. I don't think people would notice this change.
Fortunately, `zustand/context` wouldn't be much used compared to `zustand` itself,
so we can make a breaking change in v4 with the notice. Or, we create a new name in v4, if we come up with a neat idea.

Reply:
I'm less inclined to a new name, unless we can have a new name that makes sense. Let's go with this and people who will read the deprecation in release notes of v3.7 would be able to migrate and they'll have less things to migrate when v4 comes. And hey that's how users used to come to know about deprecations before the strikethrough feature, so it's not that bad :P

Reply:
That's fair. Alternatively, what do you think if we drop this type export in v4? I don't know how many people use it. Do you think of its use case? Yeah, I'm actually inclined to it, if possible. We can put it back if too people complain...

Reply:
I'm a bit reluctant but I think we could do that, even if there are use cases, users can use `typeof context.useStore`.

But I have another idea, what if we export the new type with same name from a temporary `zustand/context/forward-compat` and then mark the on exported from `zustand/context` as deprecated. In this way users will be able to migrated to the newer type and then in v4 we'll tell them to use `zustand/context` instead of `zustand/context/forward-compat`.

In fact as you review you'll find many more such types that we could put in a common forward-compat module.

And alternatively instead of deleting the forward-compat module in v4, we could even delete it in V5

Wdyt?

Reply:
Hmm, unless there's a huge concern for the migration path, I wouldn't go with `forward-compat`. I think what we have in #772 and #725 is mostly okay. It has some breaking changes, but feels reasonable.

if `typeof context.useStore` works, I would like to deprecate `export type UseContextStore` in v3.7.0 and drop(=not export) it in v4.

Reply:
Yeah also on second thoughts I too think it's okay to drop it in v4.